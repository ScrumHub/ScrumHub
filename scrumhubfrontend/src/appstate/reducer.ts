import * as Actions from "./actions";
import { createReducer, PayloadAction } from "@reduxjs/toolkit";
import { RequestResponse } from "./response";
import { IAssignBI, ILoginState, IPeopleList, IPerson, IBacklogItem, IBacklogItemList, IRepository, IRepositoryList, ISprint, ISprintList, ITask, ITaskList, IState } from "./stateInterfaces";
import { initError, loggedOutLoginState, unassignedPBI } from "./stateInitValues";
import { isArrayValid, isNameFilterValid } from "../components/utility/commonFunctions";
import { getError, updateStateTasks, updateStateKeys, addStateTask, addStateUnassignedTaskToPBI, updateStatePBI, updateTasksSWR, fetchStateRepos, updateStateOneSprint, addStateRepo, addStatePBI, addStateSprint, updateOnDragStateTasks } from "./stateUtilities";
import { isNull } from "lodash";
import config from "../configuration/config";
var _ = require('lodash');

export const reducerFunction = (initState: IState) => {
  return (createReducer(initState, {
    [Actions.clearError.type]: (state: IState, payload: any) => {
      let newState = _.cloneDeep(state as IState);
      return ({ ...newState, error: initError, changedRepo:isNameFilterValid(payload.payload)?payload.payload:""});
    },
    [Actions.clearReposList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.repositories = [];
      newState.reposRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearState.type]: (state: IState) => {
      return (initState);
    },
    [Actions.clearProject.type]: (state: IState) => {
      let newState = _.cloneDeep(state) as IState;
      return { ...initState, repositories: newState.repositories, loginState:newState.loginState, currentUser: newState.currentUser };
    },
    [Actions.clearPBIsList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.pbiPage.list = [];
      newState.productRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearSprintList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.sprintPage.list = [];
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.clearSprint.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = null;
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.updateSprintKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, sprintKeys: updateStateKeys(newState.keys.sprintKeys, temp) } };
    },
    [Actions.updatePBIKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, pbiKeys: updateStateKeys(newState.keys.pbiKeys, temp) } };
    },
    [Actions.updateSprintLoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, sprintKeys: updateStateKeys(newState.loadingKeys.sprintKeys, temp) } };
    },
    [Actions.updatePBILoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, pbiKeys: updateStateKeys(newState.loadingKeys.pbiKeys, temp) } };
    },
    [Actions.login.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const tempLogin = expanded.payload as ILoginState
      return ({ ...newState, loginState: { ...newState.loginState, isLoggedIn: tempLogin.isLoggedIn, token: tempLogin.token } });
    },
    [Actions.logout.type]: (state: IState) => { return ({ ...initState, loginState: loggedOutLoginState }); },
    /*REPOS*/
    [Actions.fetchReposThunk.pending.toString()]: (state: IState) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchReposThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepositoryList, number>>) => {
      let newState = _.cloneDeep(state);
      return fetchStateRepos(newState, payload.payload.response as IRepositoryList, _.get(payload, ["meta", "arg", "filters", "pageNumber"], config.defaultFilters.page),
       _.get(payload, ["meta", "arg", "filters", "pageSize"], config.defaultFilters.size),_.get(payload.payload.response, ["pageCount"], 1));
    },
    [Actions.fetchReposThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addRepositoryThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addRepositoryThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepository, number>>) => {
      let newState = _.cloneDeep(state);
      return (addStateRepo(newState, payload.payload.response as IRepository));
    },
    [Actions.addRepositoryThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*PBIS*/
    [Actions.finishPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.finishPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.finishPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return(addStatePBI(newState,payload.payload.response as IBacklogItem));
    },
    [Actions.addPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.deletePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.deletePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<any, any>>) => {
      let newState = _.cloneDeep(state);
      newState.loading = false;
      return newState;
    },
    [Actions.deletePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPBIsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBIsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      const pbisList = payload.payload.response as IBacklogItemList;
      if (newState.pbiPage && isArrayValid(newState.pbiPage.list) && newState.pbiPage.list[0].id === 0) {
        newState.pbiPage = { ...pbisList, list: [newState.pbiPage.list[0]].concat(pbisList.list) };
      }
      else {
        newState.pbiPage = { ...pbisList, list: [unassignedPBI].concat(pbisList.list) };
      }
      newState.loading = false;
      return newState;
    },
    [Actions.fetchPBIsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPeopleThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPeopleThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPeopleList, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, people: payload.payload.response as IPeopleList, error: initError, loading: false };
    },
    [Actions.fetchPeopleThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getCurrentUserThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getCurrentUserThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPerson, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, currentUser: payload.payload.response as IPerson, error: initError, loading: false };
    },
    [Actions.getCurrentUserThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getPBINamesThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getPBINamesThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      newState.error = initError;
      newState.namedPBI = (payload.payload.response as IBacklogItemList).list as IAssignBI[];
      newState.loading = false;
      return newState;
    },
    [Actions.getPBINamesThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.estimatePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.estimatePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.estimatePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.editPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.editPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.editPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*SPRINTS*/
    [Actions.fetchSprintsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchSprintsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprintList, number>>) => {
      let newState = _.cloneDeep(state);
      const sprints = payload.payload.response as ISprintList;
      newState.sprintPage = sprints;
      const index = sprints && isArrayValid(sprints.list) ? sprints.list.findIndex((sprint: ISprint) => sprint.isCurrent) : -1;
      if (index !== -1) { newState.activeSprintNumber = sprints.list[index].sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.loading = false;
      newState.error = initError;
      return newState;
    },
    [Actions.fetchSprintsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = payload.payload.response as ISprint;
      if (!isNull(newState.openSprint) && newState.openSprint.isCurrent) { newState.activeSprintNumber = newState.openSprint.sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.error = initError;
      newState.loading = false;
      return newState;
    },
    [Actions.fetchOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updateOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updateOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint | any, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.updateOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.completeOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.completeOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.completeOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, undefined>>) => {
      let newState = _.cloneDeep(state);
      return(addStateSprint(newState,payload.payload.response as ISprint));
    },
    [Actions.addSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*TASKS*/
    [Actions.fetchPBITasksThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBITasksThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return updateTasksSWR(newState, payload.payload.response as ITaskList);
    },
    [Actions.updateTasks.type]: (state: IState, payload: PayloadAction<any>) => {
      let newState = _.cloneDeep(state);
      const item = payload.payload as IBacklogItem;
      return updateStatePBI(newState, item);
    },
    [Actions.fetchPBITasksThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addUnassignedTasksToPBI.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addUnassignedTasksToPBI.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateUnassignedTaskToPBI(newState, payload.payload.response as ITaskList);
    },
    [Actions.addUnassignedTasksToPBI.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateTask(newState, payload.payload.response as ITask);
    },
    [Actions.addTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.assignTaskToPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.assignTaskToPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateOnDragStateTasks(newState,payload.payload.response as ITask);
    },
    [Actions.assignTaskToPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updatePersonInTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updatePersonInTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.updatePersonInTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.startTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.startTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.startTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
  }));
}