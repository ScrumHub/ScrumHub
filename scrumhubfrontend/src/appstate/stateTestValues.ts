
import config from "../configuration/config";
import { initError, initPeopleList, initPerson, initBI, initBacklogItemList, initRepository, initRepositoryList, initSprint, initSprintList, initTaskList, initTask, initKeys, initLoadingKeys } from "./stateInitValues";
import { ILoginState, IBacklogItemList, IRepositoryList, ISprintList, IState, ITaskList } from "./stateInterfaces";
import { validateUri } from "./reducerUtilities";

export const testLoginState: ILoginState = {
  isLoggedIn: false,
  token: "",
  client_id: validateUri(process.env.REACT_APP_CLIENT_ID),
  redirect_uri: validateUri(process.env.REACT_APP_REDIRECT_URI),
  client_secret: validateUri(process.env.REACT_APP_CLIENT_SECRET),
  proxy_url: validateUri(process.env.REACT_APP_PROXY_URL)
}

export const initTestState: IState = {
  loading: false,
  error: initError,
  pbiPage: initBacklogItemList,
  sprintPage: initSprintList,
  taskPage: initTaskList,
  repositories: initRepositoryList,
  openSprint: initSprint,
  reposLastPage: false,
  reposRequireRefresh: false,
  productLastPage: false,
  productRequireRefresh: false,
  sprintLastPage: false,
  sprintRequireRefresh: false,
  taskLastPage: false,
  taskRequireRefresh: false,
  tasks: [],
  namedPBI: [],
  repoId: -1,
  people: initPeopleList,
  currentUser: initPerson,
  activeSprintNumber: -1,
  keys: initKeys,
  loadingKeys: initLoadingKeys,
  loginState: testLoginState,
  changedRepo: ""
};

export const testRepositoryList: IRepositoryList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initRepository],
}

export const testPBIList: IBacklogItemList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initBI],
}

export const testSprintList: ISprintList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initSprint],
}

export const testTaskList: ITaskList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initTask],
}

export const tstConf = {
  ownerName: "ownerName",
  url: `${config.backend.ip}:${config.backend.port}/api`,
  pbiId: 0,
  sprintNr: 0,
  hours: 2,
  editedPBI: {
    "name": initBI.name,
    "priority": initBI.priority,
    "acceptanceCriteria": initBI.acceptanceCriteria
  },
  name: "taskName",
  taskId: 0,
  login: "userLogin",
  isAssign: true,
  hotfix: true
}

export const testConnectionError = {
  "code": 0,
  "response": {
    "data": null,
    "Message": "Connection error",
    "metadata": null,
    "successful": false,
  },
}
export const testFilters = { pageSize: config.defaultFilters.page, pageNumber: config.defaultFilters.size }

export const testFetchReposVals = {
  filters: testFilters,
  token: config.token,
}