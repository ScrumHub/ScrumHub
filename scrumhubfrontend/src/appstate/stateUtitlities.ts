import { IFilters } from "./stateInterfaces";

/**
 * @returns Request header
 * @param {String} token User validation
 * @param {Object} config Configuration of port and id, can be in Production or Development
 */
 export const getHeader = (token: string, config: any) => {
  return ({
    'Accept': "application/json",
    'authToken': token,
    'Access-Control-Allow-Origin': `*`,
  });
};

/**
 * @returns Request header with "application/json" content type
 * @param {String} token User validation
 * @param {Object} config Configuration of port and id, can be in Production or Development
 */
export const getHeaderWithContent = (token: string, config: any) => {
  return ({
    'authToken': token,
    'Accept': "application/json",
    'contentType': "application/json",
    'Access-Control-Allow-Origin': `*`,
  } as IFilters);
};

/**
 * @returns Request header that accepts all responses type
 * @param {String} token User validation
 * @param {Object} config Configuration of port and id, can be in Production or Development
 */
export const getHeaderAcceptAll = (token: string, config: any) => {
  return ({
    'authToken': token,
    'Accept': "*/*",
    'Access-Control-Allow-Origin': `*`,
  } as IFilters);
};

/**
 * @param {IFilters} filters Filters to validate and concatenate into string
 * @returns {String} Filters' names and values combined into one string 
 */
 export function filterUrlString(filters: IFilters) {
  return (typeof (filters) === "undefined" ? ""
    : Object.keys(filters)
      .map((filterName: string) => {
        const value = String(filters[filterName]).trim();
        return value && value !== "null" && value !== "undefined" ? `${filterName}=${value}` : "";
      })
      .filter((x) => x !== "")
      .join("&"));
};

/** Uses jest to render media object for testing 
*/
export function testMediaMatchObject(){
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
}


