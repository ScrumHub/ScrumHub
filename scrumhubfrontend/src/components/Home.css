
.reposGrid {
    display: grid;
    place-items: center;
    margin-top: 3vh;
    background: transparent;
  }
  
  .repoCard {
    background: white;
    margin-bottom: 3vh;
  }
  
  .calButton {
    width: 100%;
    overflow: hidden;
    white-space: nowrap;
  }

  .cardButton {
    width: 80%;
     /*width: 90%;*/
  }

  .ant-popover-disabled-compatible-wrapper{
      display:unset !important;
      width: 100%;
  }

  .ant-card-actions{
      padding-left: 24px !important;
      padding-right: 24px !important;
  }

