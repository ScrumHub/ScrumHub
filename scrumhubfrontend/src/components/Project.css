.projectDiv {
  margin-top: 0.25%;
  margin-bottom: 1%;
}
.projectSpace {
  transform: scale(0.96);
  margin-top: 0;
  margin-bottom: 0.25%;
  display: flex;
}
.projectDropBtn {
  color: #1890ff;
}
.projectWhIcon {
  color: white;
}
.projectColIcon {
  color: #1890ff;
}

.projectInputPpl {
  width: 160;
}

.projectCurUserDiv{
cursor: "pointer"; 
border-radius: "50%";  
}

.projectCurUserShadowDiv{
  cursor: "pointer"; 
  border-radius: "50%";  
  box-shadow:white 0px 0px 0.1px 0.05em;
  }