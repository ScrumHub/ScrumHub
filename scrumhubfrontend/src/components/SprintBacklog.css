tfoot th,
  tfoot td {
    background: #fafafa;
    font-weight: 600;
  }
  [data-theme="dark"] tfoot th,
  [data-theme="dark"] tfoot td {
    background: #1d1d1d;
  }

.summary{
    background:#fafafa;
    font-weight: 600;
}