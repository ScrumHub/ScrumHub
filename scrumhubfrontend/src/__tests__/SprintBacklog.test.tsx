/**
 * @jest-environment jsdom
 */

import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { isNull } from "lodash";
import { Provider } from "react-redux";
import { Route, BrowserRouter as Router, Routes } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import { reducerFunction } from "../appstate/reducer";
import { initTestState, updatedTestState } from "../appstate/stateTestValues";
import renderer, {create} from 'react-test-renderer';
import { Project } from "../components/Project";
import { Home } from "../components/Home";
import TestRenderer from "react-test-renderer";
import { SprintBacklog } from "../components/SprintBacklog";
import { testMediaMatchObject } from "../appstate/stateUtitlities";
const { act: actTest } = TestRenderer;


describe('Project component in container', () => {
  let container: any;
  let store: any;
  beforeEach(() => {
    testMediaMatchObject();
    container = document.createElement('div');
    document.body.appendChild(container);
    store = configureStore({
      reducer: reducerFunction(updatedTestState),
    });
  });

  afterEach(() => {
    // cleanup on exiting
    if (!isNull(container)) {
      unmountComponentAtNode(container);
      container.remove();
      container = null;
    }
  });

  it("Rendered Backlog tables", async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <Router>
            <Routes>
            <Route path={`/:owner/:name/Sprints/:number`} element={<SprintBacklog />} />
            <Route path={`/:owner/:name/active-sprint/:number`} element={<SprintBacklog />} />
              <Route path={`/`} element={<></>} />
            </Routes>
          </Router>
        </Provider>,
        container
      );
    });
    expect(
      container?.getElementsByClassName("container").length
    ).toBeGreaterThanOrEqual(0);
    expect(container.textContent).toBe("");
  });

  it('project component rendered is the same as snapshot', () => {
    const temp = renderer.create(
      <Provider store={store}>
        <Router>
          <SprintBacklog />
        </Router>
      </Provider>).toJSON();
    expect(temp).toMatchSnapshot();
  });

  it("Submitting a name via the input field changes the name state value", () => {
    actTest(() => { create(<Provider store={store}>
      <Router>
        <Routes>
        <Route path={`/:owner/:name/Sprints/:number`} element={<SprintBacklog />} />
            <Route path={`/:owner/:name//:number`} element={<SprintBacklog />} />
              <Route path={`/`} element={<></>} />
        </Routes>
      </Router>
    </Provider>)});
  });
  it("No Expanded rows", () => {
    store = configureStore({
      reducer: reducerFunction(initTestState),
    });
    actTest(() => {create(<Provider store={store}>
      <Router>
        <Routes>
          <Route path={`/:owner/:name`} element={<Project />} />
          <Route path={`/`} element={<Home />} />
        </Routes>
      </Router>
    </Provider>)});
  });
  it("Rendered backlog tables", async () => {
    store = configureStore({
      reducer: reducerFunction(initTestState),
    });
    await act(async () => {
      render(
        <Provider store={store}>
          <Router>
            <Routes>
              <Route path={`/:owner/:name`} element={<Project />} />
              <Route path={`/`} element={<Home />} />
            </Routes>
          </Router>
        </Provider>,
        container
      );
    });
    expect(
      container?.getElementsByClassName("container").length
    ).toBeGreaterThanOrEqual(0);
    expect(container.textContent).toBe("");
  });
  it('project component loading', async () => {
    store = configureStore({
      reducer: reducerFunction({...initTestState, loading:true}),
    });
    await act(async () => {
      render(
        <Provider store={store}>
          <Router>
            <Routes>
              <Route path={`/:owner/:name`} element={<Project />} />
              <Route path={`/`} element={<Home />} />
            </Routes>
          </Router>
        </Provider>,
        container
      );
    });
    expect(
      container?.getElementsByClassName("container").length
    ).toBeGreaterThanOrEqual(0);
    expect(container.textContent).toBe("");
  });
});
