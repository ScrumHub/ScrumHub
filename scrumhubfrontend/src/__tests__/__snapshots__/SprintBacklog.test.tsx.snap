// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project component in container project component rendered is the same as snapshot 1`] = `
<div
  style={
    Object {
      "marginBottom": "1%",
      "marginLeft": "2%",
      "marginRight": "2%",
      "marginTop": 0,
    }
  }
>
  <section
    className="card"
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <section
    className="card"
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <div
    className="ant-space ant-space-horizontal ant-space-align-center"
    style={Object {}}
  >
    <div
      className="ant-space-item"
      style={Object {}}
    >
      <article
        className="ant-typography"
      >
        
      </article>
    </div>
  </div>
  <br />
  <div
    className="ant-space ant-space-horizontal ant-space-align-center"
    style={Object {}}
  >
    <div
      className="ant-space-item"
      style={Object {}}
    >
      <button
        className="ant-btn ant-btn-link"
        onClick={[Function]}
        type="button"
      />
    </div>
  </div>
</div>
`;
